File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/SlowSessionHandler.java
package systems.zlink.e2e.spotservice.shared;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkTypedSessionPacketHandler;

public final class SlowSessionHandler
    implements ZLinkTypedSessionPacketHandler<ZLinkSessionContext, Contracts.SlowSessionReq> {
    private final ScenarioState evidence;

    public SlowSessionHandler(ScenarioState evidence) {
        this.evidence = evidence;
    }

    @Override
    public Class<Contracts.SlowSessionReq> messageType() {
        return Contracts.SlowSessionReq.class;
    }

    @Override
    public CompletionStage<Void> handle(
        ZLinkSessionContext context,
        ZLinkSessionDispatchContext dispatch,
        Contracts.SlowSessionReq request) {
        try {
            Thread.sleep(Math.max(0, request.delayMilliseconds()));
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            evidence.record("SlowSessionInterrupted", "session", request.value());
        }
        evidence.record("SlowSessionHandled", "session", request.value());
        try {
            context.client().reply(new Contracts.SlowSessionRes(request.value())).submit();
        } catch (Exception error) {
            evidence.record("SlowSessionReplyFailed", "session", request.value() + "/" + error.getClass().getSimpleName());
        }
        return CompletableFuture.completedFuture(null);
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Session/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/session/handlers/SlowSessionHandler.kt
package systems.zlink.e2e.kotlin.spotservice.session.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.ScenarioState
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import kotlinx.coroutines.delay
import systems.zlink.framework.kotlin.ZLinkSuspendingTypedSessionPacketHandler

class SlowSessionHandler(
    private val evidence: ScenarioState
) : ZLinkSuspendingTypedSessionPacketHandler<ZLinkSessionContext, Contracts.SlowSessionReq> {
    override fun packetName(): String = "SlowSessionReq"

    override fun messageType(): Class<Contracts.SlowSessionReq> = Contracts.SlowSessionReq::class.java

    override suspend fun handle(
        context: ZLinkSessionContext,
        dispatch: ZLinkSessionDispatchContext,
        request: Contracts.SlowSessionReq
    ) {
        delay(request.delayMilliseconds.coerceAtLeast(0).toLong())
        evidence.record("SlowSessionHandled", "session", request.value)
        try {
            context.client()
                .reply(Contracts.SlowSessionRes(request.value))
                .submit()
        } catch (error: Exception) {
            evidence.record("SlowSessionReplyFailed", "session", "${request.value}/${error.javaClass.simpleName}")
        }
    }
}
